sphinx/directives/code.py

```diff
@@ -156,6 +154,190 @@ def run(self):
+            if option1 in self.options and option2 in self.options:
+                raise ValueError(_('Cannot use both "%s" and "%s" options') %
+                                 (option1, option2))
+
+    def read_file(self, filename):
+        # type: (unicode) -> List[unicode]
+        try:
+            with codecs.open(filename, 'r', self.encoding, errors='strict') as f:  # type: ignore  # NOQA
+                text = f.read()  # type: unicode
+                if 'tab-width' in self.options:
+                    text = text.expandtabs(self.options['tab-width'])
+
+                lines = text.splitlines(True)
+                if 'dedent' in self.options:
+                    return dedent_lines(lines, self.options.get('dedent'))
+                else:
+                    return lines
+        except (IOError, OSError) as exc:
+            raise IOError(_('Include file %r not found or reading it failed') % filename)
+        except UnicodeError as exc:
+            raise UnicodeError(_('Encoding %r used for reading included file %r seems to '
+                                 'be wrong, try giving an :encoding: option') %
+                               (self.encoding, filename))
+
+    def read(self):
+        # type: () -> Tuple[unicode, int]
+        if 'diff' in self.options:
+            lines = self.show_diff()
+        else:
+            filters = [self.pyobject_filter,
+                       self.lines_filter,
+                       self.start_filter,
+                       self.end_filter,
+                       self.prepend_filter,
+                       self.append_filter]
+            lines = self.read_file(self.filename)
+            for func in filters:
+                lines = func(lines)
+
+        return ''.join(lines), len(lines)
+
+    def show_diff(self):
+        # type: () -> List[unicode]
+        new_lines = self.read_file(self.filename)
+        old_filename = self.options.get('diff')
+        old_lines = self.read_file(old_filename)
+        diff = unified_diff(old_lines, new_lines, old_filename, self.filename)  # type: ignore
+        return list(diff)
+
+    def pyobject_filter(self, lines):
+        # type: (List[unicode]) -> List[unicode]
+        pyobject = self.options.get('pyobject')
+        if pyobject:
+            from sphinx.pycode import ModuleAnalyzer
+            analyzer = ModuleAnalyzer.for_file(self.filename, '')
+            tags = analyzer.find_tags()
+            if pyobject not in tags:
+                raise ValueError(_('Object named %r not found in include file %r') %
+                                 (pyobject, self.filename))
+            else:
+                start = tags[pyobject][1]
+                end = tags[pyobject][2]
+                lines = lines[start - 1:end - 1]
+                if 'lineno-match' in self.options:
+                    self.lineno_start = start
+
+        return lines
+
+    def lines_filter(self, lines):
```

Comment:
I wonder how the `:lines: 1,1,1,1` option works. Is it invalid?
Now it is allowed and shows the first line 4 times.